godal.go

```diff
@@ -4060,6 +4060,88 @@ func (ds *Dataset) Dem(destPath, processingMode string, colorFilename string, sw
 }
 
+// ViewshedMode is the "cell height calculation mode" for the viewshed process
+//
+// Source: https://github.com/OSGeo/gdal/blob/master/alg/viewshed/viewshed_types.h
+type ViewshedMode uint32
+
+const (
+	// MDiagonal is the "diagonal mode"
+	MDiagonal ViewshedMode = iota + 1
+	// MEdge is the "edge mode"
+	MEdge
+	// MMax is the "maximum value produced by Diagonal and Edge mode"
+	MMax
+	// MMin is the "minimum value produced by Diagonal and Edge mode"
+	MMin
+)
+
+// ViewshedOutputType sets the return type and information represented by the returned data
+//
+// Source: https://gdal.org/en/stable/programs/gdal_viewshed.html
+//
+// NOTE: "Cumulative (ACCUM)" mode not currently supported, as it's not available in the `GDALViewshedGenerate` function
+// (it's only used in the command line invocation of `viewshed`)
+type ViewshedOutputType uint32
+
+const (
+	// Normal returns a raster of type Byte containing visible locations
+	Normal ViewshedOutputType = iota + 1
+	// MinTargetHeightFromDem return a raster of type Float64 containing the minimum target height for target to be visible from the DEM surface
+	MinTargetHeightFromDem
+	// MinTargetHeightFromGround return a raster of type Float64 containing the minimum target height for target to be visible from ground level
+	MinTargetHeightFromGround
+)
+
+// Viewshed (binding for GDALViewshedGenerate), creates a viewshed from a raster DEM, these parameters (mostly) map to to parameters for GDALViewshedGenerate
+//
+// for more information see: https://gdal.org/en/stable/api/gdal_alg.html#_CPPv420GDALViewshedGenerate15GDALRasterBandHPKcPKc12CSLConstListddddddddd16GDALViewshedModed16GDALProgressFuncPv22GDALViewshedOutputType12CSLConstList
+//
+// Creations options can be set through options with:
+//
+//	Viewshed(bnd, "mem", "none", ..., CreationOption("TILED=YES","BLOCKXSIZE=256"))
+func Viewshed(targetBand Band, driverName *DriverName, targetRasterName string, observerX float64, observerY float64, observerHeight float64, targetHeight float64,
+	visibleVal float64, invisibleVal float64, outOfRangeVal float64, noDataVal float64, curveCoeff float64, mode ViewshedMode, maxDistance float64,
+	heightMode ViewshedOutputType, opts ...ViewshedOption) (*Dataset, error) {
+
+	// TODO: Should I put a 'warning' in the documentation for `Viewshed` instead of disallowing the last two configurations?
+	if !CheckMinVersion(3, 1, 0) {
+		return nil, errors.New("failed to run, 'viewshed' not supported on GDAL versions < 3.1.0")
+	} else if !CheckMinVersion(3, 4, 2) {
+		return nil, errors.New("cannot run 'viewshed' with GDAL version <= 3.4.1, as some tests produce invalid results under these conditions")
+	} else if !CheckMinVersion(3, 10, 0) && heightMode == MinTargetHeightFromDem {
+		return nil, errors.New("height mode CANNOT be `MinTargetHeightFromDem` when running a GDAL version < 3.10, as some tests produce invalid results under these conditions")
+	}
```

Comment:
I would just leave this section out, it is not godal's role to circumvent fixed bugs in gdal. A comment with a link to the gdal PR/issue numbers is sufficient I believe